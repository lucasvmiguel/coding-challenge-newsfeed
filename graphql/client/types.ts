export enum resolveTypes {
  User = 'User',
  Announcement = 'Announcement',
  Project = 'Project'
}
