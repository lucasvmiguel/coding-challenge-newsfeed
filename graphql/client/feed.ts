import { gql } from '@apollo/client'

export const FEED_QUERY = gql`
  query Feed($user_type: Fellowship!, $limit: Int, $offset: Int) {
    feed(user_type: $user_type, limit: $limit, offset: $offset) {
      __typename
      ... on User {
        id
        name
        bio
        avatar_url
        fellowship
        created_ts
        projects {
          id
          name
          icon_url
        }
      }
      ... on Announcement {
        id
        title
        body
        fellowship
        created_ts
      }
      ... on Project {
        id
        name
        description
        icon_url
        created_ts
        users {
          id
          name
          avatar_url
        }
      }
    }
  }
`
