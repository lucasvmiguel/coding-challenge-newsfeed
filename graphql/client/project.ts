import { gql } from '@apollo/client'

export const PROJECT_QUERY = gql`
  query project($id: Int!) {
    project(id: $id) {
      id
      name
      description
      icon_url
      users {
        id
        name
        avatar_url
      }
    }
  }
`
