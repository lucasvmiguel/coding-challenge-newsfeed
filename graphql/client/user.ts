import { gql } from '@apollo/client'

export const USER_QUERY = gql`
  query user($id: Int!) {
    user(id: $id) {
      id
      name
      bio
      fellowship
      avatar_url
      projects {
        id
        name
        icon_url
      }
    }
  }
`