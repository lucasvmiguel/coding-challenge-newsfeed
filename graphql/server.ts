import { ApolloServer, gql } from 'apollo-server-micro'

import { resolveTypes } from './client/types';
import * as resolvers from './resolvers'


const typeDefs = gql`
  enum Fellowship {
    founders
    angels
    writers
    all
  }

  type Project {
    id: Int!
    name: String!
    description: String!
    icon_url: String!
    users: [User!]!
    created_ts: String!
  }

  type User {
    id: Int!
    name: String!
    bio: String!
    avatar_url: String!
    fellowship: Fellowship!
    projects: [Project!]!
    created_ts: String!
  }

  type Announcement{
    id: Int!
    fellowship: Fellowship!
    title: String!
    body: String!
    created_ts: String!
    updated_ts: String!
  }

  union FeedItem = User | Announcement | Project

  type Query {
    project(id: Int!): Project!
    user(id: Int!): User!
    feed(user_type: Fellowship!, limit: Int, offset: Int): [FeedItem!]!
  }
`;

export const server = new ApolloServer({
  typeDefs,
  resolvers: {
    ...resolvers,
    FeedItem: {
      __resolveType(obj: any, context: any, info: any) {
        if (obj.bio !== null) {
          return resolveTypes.User;
        }

        if (obj.body !== null) {
          return resolveTypes.Announcement;
        }

        if (obj.description !== null) {
          return resolveTypes.Project;
        }

        return null;
      },
    },
  }
})
