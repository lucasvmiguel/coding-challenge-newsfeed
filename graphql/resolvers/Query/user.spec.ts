import { USER_QUERY } from '../../client/user'
import { server } from '../../server'

it('fetches user', async () => {
  const res = await server.executeOperation({ query: USER_QUERY, variables: { id: 1 } })
  expect(res).toMatchSnapshot()
})
