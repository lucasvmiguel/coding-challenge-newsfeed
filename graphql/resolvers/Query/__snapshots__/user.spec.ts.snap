// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "avatar_url": "https://avatars.dicebear.com/api/avataaars/a1.svg",
      "bio": "I'm a founder, angel investor, ex-google team member passionate about startups. I also love traveling, the digital nomad lifestyle and anything outdoors.

I'm currently exploring several ideas, excited to launch a new concept or join an existing team.",
      "fellowship": "founders",
      "id": 1,
      "name": "Dennis Pearson",
      "projects": Array [
        Object {
          "icon_url": "https://avatars.dicebear.com/api/jdenticon/a1.svg",
          "id": 1,
          "name": "Seam",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
